### ISSUE TITLE
Docstring Style

### ISSUE THREAD

**Issue by dbrgn** [py3minepi/py3minepi-legacy#33]
What docstring style should we use?

My currently favorite style is:

```
"""
Start on second line, provide a short description of the entire code.

Add more details here. This may contain examples, links, explanations, reasons, etc.

Document the arguments here. End with an empty line.

"""
```

And:

```
"""
If the docstring is a one-liner, leave away the trailing empty line.
"""
```

For the docstring style we could use Sphinx syntax to be able to generate parseable documentation.

**Comment by doismellburning** [comment]
fwiw my favourite is "whichever some highly-regarded Python linty thing uses" /cc @ghickman

**Comment by hashbangstudio** [comment]
"whichever some highly-regarded Python linty thing uses"  :+1:

**Comment by dbrgn** [comment]
There's no clear standard, as far as I know. The PEP regarding docstrings does not describe all the details.

**Comment by doismellburning** [comment]
:(

**Comment by doismellburning** [comment]
@ghickman-as-a-Service? ;P

**Comment by ghickman** [comment]
[PEP257](http://legacy.python.org/dev/peps/pep-0257/#one-line-docstrings) should cover everything.

One line:

```
"""Simple one line docstring."""
```

Multi-line:

```
"""
Short summary like a subject line.

More in depth body of the docstring, giving more information and possibly
even a basic example.
"""
```

Ideally docstrings are limited to 78 characters since they are prose but I don't feel this needs to be a hard limit.

**Comment by doismellburning** [comment]
Yay PEP257. Do you have an automated enforcer handy?

**Comment by ghickman** [comment]
I don't unfortunately…

**Comment by dbrgn** [comment]
PEP257 is vague on various things. E.g. it specified that single-line docstrings are possible, but not that they can't stand on their own line, or which one is better. Rather it explains what a docstring is, what it should contain and how it should be parsed.

I'd vote for simply showing a few good examples. On the one line docstring thing I don't have a big preference, but usually I'd go for consistency (always starting on the second line). Also makes conversion from one style to the other unnecessary once someone adds more content to the docstring.

**Comment by ghickman** [comment]
> One-liners are for really obvious cases. They should really fit on one line.

This seems pretty simple to me ;)

But I would agree with everything you say nonetheless, especially putting the single line docstring on a newline. Clean diffs are always better than the alternative.

For multiline docstrings I find Tim Pope's [guide](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) to git commits can be very useful on how to separate the "subject" and "body".

How's this?

One line:

```
"""
Simple one line docstring.
"""
```

Multi-line:

```
"""
Short summary like a subject line.

More in depth body of the docstring, giving more information and possibly
even a basic example.
"""
```

**Comment by doismellburning** [comment]
+1

**Comment by dbrgn** [comment]
Yes, that part is obvious. The question is whether we want a trailing empty line (useful to separate the docstring visually from the following function definition, as according to some PEP IIRC the code should follow the `"""` directly), or whether to use Sphinx style argument docs (and other directives).

**Comment by ghickman** [comment]
Personally I prefer no line between the docstring text and the `"""` and a blank line after. However this is a personal preference so if there was overwhelming consensus another way I'd be happy with that.

**Comment by doismellburning** [comment]
@dbrgn Can you clarify what "[t]he question is" please? I've revisited this issue trying to understand where we're at, and I confess to having no idea what's actually being debated...

**Comment by dbrgn** [comment]
### Format / Whitespace:

```
"""
Short summary like a subject line.

More in depth body of the docstring, giving more information and possibly
even a basic example.
"""

code_starts_here = 1
```

vs

```
"""
Short summary like a subject line.

More in depth body of the docstring, giving more information and possibly
even a basic example.

"""
code_starts_here = 1
```

I'd vote for the latter.

### Single line docstrings

```
"""Short docstring."""
```

vs

```
"""Short docstring.
"""
```

vs

```
"""
Short docstring.
"""
```

I'd vote for the last example. If a docstring is extended, we have cleaner git diffs.

### Argument docs

```
"""
Sphinx-style docstring.

More in depth body of the docstring, giving more information and possibly
even a basic example.

:param height: The height of the block.
:type height: int
:param width: The width of the block.
:type width: int
:returns: A special type instance.
:rtype: :class:`package.module.SpecialType`

"""
```

vs

```
"""
Google-style docstring.

More in depth body of the docstring, giving more information and possibly
even a basic example.

Args:
    height (int):
        The height of the block.
    width (int):
        The width of the block.

Returns:
    A ``ackage.module.SpecialType`` instance.

"""
```

vs some other formats like Epydoc.

For non-library software I usually prefer the Google-Style docstrings, as they're easily readable. But for libraries I'd go with Sphinx-Style, so we could auto-generate API docs.

**Comment by doismellburning** [comment]
@dbrgn <3 thanks for clarifying!

**Comment by doismellburning** [comment]
(1/3/same-as-you here)

**Comment by pozorvlak** [comment]
1, 1 or 3, don't care here.

**Comment by ghickman** [comment]
I've been pointed to a docstring checking tool, [pep257](https://pypi.python.org/pypi/pep257) and a [flake8 wrapper](https://pypi.python.org/pypi/flake8-docstrings) for it (thanks @meshy!).

Some error codes I'm not sure about:

D100: seems overkill to do this for _every_ module.
D200: I still agree with your point earlier so torn here.
D20{3,4}: overkill on whitespace in classes to me.

re examples: 1, 1/3, don't mind

**Comment by doismellburning** [comment]
@ghickman Whack it in a PR with vanilla settings, see how it goes?

**Comment by meshy** [comment]
Pleasure `:)`

**Comment by dbrgn** [comment]
Wow, pep257 is nice :)
